agent/src/com/cloud/agent/direct/download/HttpDirectTemplateDownloader.java

```diff
@@ -0,0 +1,158 @@
+        File f = new File(getDestPoolPath() + File.separator + downloadDir + File.separator + getFileNameFromUrl());
+        setDownloadedFilePath(f.getAbsolutePath());
+    }
+
+    private GetMethod createRequest(String downloadUrl, Map<String, String> headers) {
+        GetMethod request = new GetMethod(downloadUrl);
+        request.getParams().setParameter(HttpMethodParams.RETRY_HANDLER, myretryhandler);
+        request.setFollowRedirects(true);
+        for (String key : headers.keySet()) {
+            request.setRequestHeader(key, headers.get(key));
+        }
+        return request;
+    }
+
+    private long getRemoteSize(GetMethod request) {
+        Header contentLengthHeader = request.getResponseHeader("Content-Length");
+        long reportedRemoteSize = 0;
+        if (contentLengthHeader == null) {
+            Header chunkedHeader = request.getResponseHeader("Transfer-Encoding");
+            if (chunkedHeader == null || !"chunked".equalsIgnoreCase(chunkedHeader.getValue())) {
+                throw new CloudRuntimeException("Error template remote size");
+            }
+        } else {
+            reportedRemoteSize = Long.parseLong(contentLengthHeader.getValue());
+        }
+        return reportedRemoteSize;
+    }
+
+    private HttpMethodRetryHandler createRetryTwiceHandler() {
+        return new HttpMethodRetryHandler() {
+            @Override
+            public boolean retryMethod(final HttpMethod method, final IOException exception, int executionCount) {
+                if (executionCount >= 2) {
+                    // Do not retry if over max retry count
+                    return false;
+                }
+                if (exception instanceof NoHttpResponseException) {
+                    // Retry if the server dropped connection on us
+                    return true;
+                }
+                if (!method.isRequestSent()) {
+                    // Retry if the request has not been sent fully or
+                    // if it's OK to retry methods that have been sent
+                    return true;
+                }
+                // otherwise do not retry
+                return false;
+            }
+        };
+    }
+
+    @Override
+    public boolean downloadTemplate() {
+        InputStream in = null;
+        RandomAccessFile out = null;
+        try {
+            client.executeMethod(request);
+            remoteSize = getRemoteSize(request);
+            in = request.getResponseBodyAsStream();
+            File f = new File(getDownloadedFilePath());
+            out = new RandomAccessFile(f, "rw");
+            out.seek(0);
+            copyBytes(in, out);
+        } catch (IOException e) {
+            s_logger.error("Error downloading template " + getTemplateId() +  e.getMessage());
+            return false;
+        } finally {
+            try {
+                out.close();
+                in.close();
```

Comment:
Would be nicer to move the code in a try with resource statement (https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)